app/models/report.rb

```diff
@@ -134,4 +134,14 @@ def not_wip_previous_of_user
           .order(reported_on: :desc)
           .second
   end
+
+  def save_with_uniqueness_check
+    ActiveRecord::Base.transaction do
```

Comment:
[must]transactionは一連の操作を「全成功」か「全失敗」させることにより、矛盾を防ぐものです。
そのため、失敗した場合にはRollbackしないといけませんがこのメソッドではコミットされてしまっています。（内部で何も行なっていないのでDB上は違いはありませんが、期待される結果ではありません）
Rollbackされるように修正をお願いいたします！

Reply:
元のコードを見るとRollbackされていないことを確認しました。
こちら

```ruby
def save_with_uniqueness_check
  ActiveRecord::Base.transaction do
    if Report.where(user_id:, reported_on:).exists?
      errors.add(:reported_on, 'はすでに存在します')
      raise ActiveRecord::Rollback
    end
    save
  end
end
```

と明示的にRollbackするように修正いたしました。

Reply:
ありがとうございます！
こちらも確認したところRollbackされていませんでした。
``` ruby
  def save_with_uniqueness_check
    ActiveRecord::Base.transaction do
      if Report.where(user_id:, reported_on:).exists?
        errors.add(:reported_on, 'はすでに存在します')
        raise ActiveRecord::Rollback
      end
      raise ActiveRecord::Rollback unless save

      true
    end
  end
  ```
  としてsaveされない場合はRollbackするように修正しました。
  
  同じタイトルが存在した場合：

```shell
report2 = Report.new(
  user: user,
  reported_on: '2022-06-12',
  title: 'test55',
  description: 'test',
  emotion: 'happy'
)
=> 
#<Report:0x00007f0dcf0f4e50
...
if report2.save_with_uniqueness_check
  puts "レポート2: 保存"
else
  puts "レポート2: エラー #{report2.errors.full_messages.join(', ')}"
end
  TRANSACTION (0.6ms)  BEGIN
  Report Exists? (1.5ms)  SELECT 1 AS one FROM "reports" WHERE "reports"."user_id" = $1 AND "reports"."reported_on" = $2 LIMIT $3  [["user_id", 253826460], ["reported_on", "2022-06-12"], ["LIMIT", 1]]
  Report Exists? (1.0ms)  SELECT 1 AS one FROM "reports" WHERE "reports"."title" = $1 AND "reports"."user_id" = $2 LIMIT $3  [["title", "test55"], ["user_id", 253826460], ["LIMIT", 1]]
  Report Exists? (0.9ms)  SELECT 1 AS one FROM "reports" WHERE "reports"."reported_on" = $1 AND "reports"."user_id" = $2 LIMIT $3  [["reported_on", "2022-06-12"], ["user_id", 253826460], ["LIMIT", 1]]
  TRANSACTION (0.6ms)  ROLLBACK
レポート2: エラー タイトルはすでに存在します
=> nil
```

また、
```ruby
raise ActiveRecord::Rollback unless save
```
あとの`true`なのですが、`true`返さない形でもデータの保存自体は問題なく行われていました。しかし、以下のコードの実行結果では、保存が成功しているにもかかわらず、elseの`レポート2: エラー`が表示されてしまいました。
```shell
if report2.save_with_uniqueness_check
  puts "レポート2: 保存"
else
  puts "レポート2: エラー #{report2.errors.full_messages.join(', ')}"
end
  TRANSACTION (0.4ms)  BEGIN
  Report Exists? (1.4ms)  SELECT 1 AS one FROM "reports" WHERE "reports"."user_id" = $1 AND "reports"."reported_on" = $2 LIMIT $3  [["user_id", 253826460], ["reported_on", "2022-06-05"], ["LIMIT", 1]]
  Report Exists? (1.0ms)  SELECT 1 AS one FROM "reports" WHERE "reports"."title" = $1 AND "reports"."user_id" = $2 LIMIT $3  [["title", "test123"], ["user_id", 253826460], ["LIMIT", 1]]
  Report Exists? (0.7ms)  SELECT 1 AS one FROM "reports" WHERE "reports"."reported_on" = $1 AND "reports"."user_id" = $2 LIMIT $3  [["reported_on", "2022-06-05"], ["user_id", 253826460], ["LIMIT", 1]]
  Report Create (0.9ms)  INSERT INTO "reports" ("user_id", "title", "description", "created_at", "updated_at", "reported_on", "emotion") VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING "id"  [["user_id", 253826460], ["title", "test123"], ["description", "test"], ["created_at", "2024-08-19 02:44:09.909464"], ["updated_at", "2024-08-19 02:44:09.909464"], ["reported_on", "2022-06-05"], ["emotion", 2]]
  Watch Exists? (1.3ms)  SELECT 1 AS one FROM "watches" WHERE "watches"."user_id" = $1 AND "watches"."watchable_id" = $2 AND "watches"."watchable_type" = $3 LIMIT $4  [["user_id", 253826460], ["watchable_id", 1066585839], ["watchable_type", "Report"], ["LIMIT", 1]]
  Watch Create (0.9ms)  INSERT INTO "watches" ("watchable_type", "watchable_id", "created_at", "updated_at", "user_id") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["watchable_type", "Report"], ["watchable_id", 1066585839], ["created_at", "2024-08-19 02:44:09.938174"], ["updated_at", "2024-08-19 02:44:09.938174"], ["user_id", 253826460]]
  User Update (1.1ms)  UPDATE "users" SET "updated_at" = $1 WHERE "users"."id" = $2  [["updated_at", "2024-08-19 02:44:09.940861"], ["id", 253826460]]
  TRANSACTION (15.1ms)  COMMIT
レポート2: エラー 
=> nil
```
これを修正するために正常に保存が完了した場合に`true`を返すようにしました。

Reply:
[ask]`save!`を使うのではなく、明示的にtrueを返すようにした理由は何かありますか？
```suggestion
    ActiveRecord::Base.transaction do
      if Report.where(user_id:, reported_on:).exists?
        errors.add(:reported_on, 'はすでに存在します')
        raise ActiveRecord::Rollback
      end
      save!
    end
```

Reply:
ご指摘ありがとうございます。
>トランザクションは例外をキャッチしたら自動でRollbackを行なってくれます。

そういうのを知らずに今のコードからどうすれば解決できるかだけでコードを修正していました。
きちんと確認するように気を付けます。
参考のURLもありがとうございます。

Reply:
@reckyy 

お疲れ様です。
`save_with_uniqueness_check`メソッドを以下のように`save!`に修正した際の挙動についてなのですが、
```ruby
ActiveRecord::Base.transaction do
    if Report.where(user_id:, reported_on:).exists?
      errors.add(:reported_on, 'はすでに存在します')
      raise ActiveRecord::Rollback
    end
    save!
end
```

`app/controllers/reports_controller.rb`の`create`メソッドが `if @report.save_with_uniqueness_check`で条件分岐をしています。
```ruby
def create
  @report = Report.new(report_params)
  @report.user = current_user
  set_wip
  canonicalize_learning_times(@report)

  if @report.save_with_uniqueness_check
    Newspaper.publish(:report_save, { report: @report })
    redirect_to redirect_url(@report), notice: notice_message(@report), flash: flash_contents(@report)
  else
    render :new
  end
end
```
この場合`save_with_uniqueness_check`メソッドが`true`または`false`を返すことが期待されています。
しかし、`save!`を使って例外を発生させた場合、トランザクションはロールバックされますが、`false`は返さないようです。
その結果、
```ruby
if @report.save_with_uniqueness_check
```
の部分が処理することができず、エラーになってしまうことを確認しました。

そこで次にどう修正をしていくかなのですが、
save! を使わずに元のコードに戻す方が良いか。
引き続きsave!を使って他を修正する方が良いか。
どちらの方法が良いか、ご意見いただけますと幸いです。

Reply:
@su-su-su-su 
お疲れ様です。

createメソッドの分岐、失念していました。。申し訳ないです😭
となると、unless文でRollbackさせていたコードが良いかもです。
不適切な提案してしまい、申し訳ないです。

Reply:
@reckyy 
トランザクションで`save!`を使う意味なども知れたのでむしろ勉強になりました🙇
ありがとうございます！
こちら再度コードを戻すように修正しました。
再度ご確認いただけますでしょうか。
よろしくお願いいたします。

Reply:
ありがとうございます！
確認しました〜。

Reply:
ありがとうございます🙇‍♂️